### PR TITLE
fix: bug with subkeys, verify test fields on startup

### LINKED ISSUE #2348: If the user deletes or changes a key in the DB, all previous encrypted rows will error when accessing
Deleting a row is easy for the user to do without realising the implications.

However updating the row required a fair amount of effort so maybe less important.

Solution A

FK from each encrypted row into the keys table, so that the rows must be deleted before deleting the key. This won't help with a user modifying the key.

Solution B

Have some data in the DB that is encrypted and checked on startup. If it can't decrypt we know the key was deleted or changed.

### LINKED ISSUE #2466: Race condition related to encryption: concurrent map writes, getDerivedPrimitive
https://github.com/TBD54566975/ftl/actions/runs/10479096401/attempts/1

```
    harness.go:327: debug: fatal error: concurrent map writes
    harness.go:327: debug: 
    harness.go:327: debug: goroutine 2481 [running]:
    harness.go:327: debug: github.com/TBD54566975/ftl/internal/encryption.(*KMSEncryptor).getDerivedPrimitive(0xc0005a31c0, {0x7fd7242d3b78, 0xc00133f608})
```

### PR DESCRIPTION
**Pull request by gak**
Fixes #2348
Fixes #2466

### CODE REVIEW

File: backend/controller/dal/encryption.go [hunk 1]
```diff
@@ -84,5 +88,80 @@ func (d *DAL) EnsureKey(ctx context.Context, generateKey func() ([]byte, error))
...
+	if err != nil {
+		return fmt.Errorf("failed to update encryption verification: %w", err)
+	}
+
+	return nil
+}
+
+// verifySubkey checks if the subkey is set and if not, sets it to a verification string.
+// It returns true if the subkey was set.
+func verifySubkey[SK encryption.SubKey](encryptor encryption.DataEncryptor, encrypted *optional.Option[encryption.EncryptedColumn[SK]]) (bool, error) {
```
Comment (stuartwdouglas) [inline backend/controller/dal/encryption.go:146]: This is a bit of a nit, but passing around pointers to Option feels kinda wrong to me. Why not just return the resulting key?

**Review by stuartwdouglas — APPROVED** [review]

File: backend/controller/dal/encryption.go (hunk 1 above)
Comment (gak) [reply backend/controller/dal/encryption.go:146]: Good point. It was just the way I wrote it initially and tunnel visioned it. Will fix!

File: backend/controller/dal/encryption.go [hunk 2]
```diff
@@ -84,5 +88,86 @@ func (d *DAL) EnsureKey(ctx context.Context, generateKey func() ([]byte, error))
...
+
+func (d *DAL) verifyEncryptor(ctx context.Context) (err error) {
+	var tx *Tx
+	tx, err = d.Begin(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to begin transaction: %w", err)
+	}
+	defer tx.CommitOrRollback(ctx, &err)
+
+	var row sql.GetOnlyEncryptionKeyRow
```
Comment (alecthomas) [inline backend/controller/dal/encryption.go:105]: No need to do this, Go will reuse the existing `err` if you do `row, err := ...`

File: backend/controller/dal/encryption.go [hunk 3]
```diff
@@ -84,5 +88,86 @@ func (d *DAL) EnsureKey(ctx context.Context, generateKey func() ([]byte, error))
...
+	if err != nil {
+		return fmt.Errorf("failed to update encryption verification: %w", err)
+	}
+
+	return nil
+}
+
+// verifySubkey checks if the subkey is set and if not, sets it to a verification string.
+// returns (nil, nil) if verified and not changed
+func verifySubkey[SK encryption.SubKey](encryptor encryption.DataEncryptor, encrypted optional.Option[encryption.EncryptedColumn[SK]]) (*encryption.EncryptedColumn[SK], error) {
```
Comment (alecthomas) [inline backend/controller/dal/encryption.go:152]: `encryption.EncryptedColumn` is a `[]byte` so `nil` will already work with it without having to make it a pointer.

File: backend/controller/sql/schema/20240820011612_encryption_verification.sql [hunk 4]
```diff
@@ -0,0 +1,11 @@
+-- migrate:up
+
+ALTER TABLE encryption_keys
+    ADD COLUMN verify_timeline encrypted_timeline,
+    ADD COLUMN verify_async encrypted_async;
+
+-- migrate:down
+
+ALTER TABLE encryption_keys
```
Comment (alecthomas) [inline backend/controller/sql/schema/20240820011612_encryption_verification.sql:9]: We don't support down, so this isn't necessary really.

**Review by alecthomas — APPROVED** [review]
Nice!

**Comment by gak** [comment]
Smoke test from an older version 👌

File: internal/encryption/encryption.go [hunk 5]
```diff
@@ -206,7 +208,10 @@ func deriveKeyset(root keyset.Handle, salt []byte) (*keyset.Handle, error) {
 }
 
 func (k *KMSEncryptor) getDerivedPrimitive(subKey SubKey) (tink.AEAD, error) {
-	if primitive, ok := k.cachedDerived[subKey]; ok {
+	k.cachedDerivedMu.RLock()
+	primitive, ok := k.cachedDerived[subKey]
+	k.cachedDerivedMu.RUnlock()
```
Comment (alecthomas) [inline internal/encryption/encryption.go:213]: I'd just use a Mutex and do:

```go
k.cachedDerivedMu.Lock()
defer k.cachedDerivedMu.Unlock()
```

**Review by alecthomas — APPROVED** [review]

**Pull request merged**